[Plugins] Build and install `_CompilerPluginSupport` module directory.

DougGregor opened the pull request:
Fix the generated layout for the `_CompilerPluginSupport` Swift module, so that properly-named Swift interfaces are placed within a `.swiftmodule` subdirectory. Install that directory into lib/swift.

DougGregor committed:
[Plugins] Build and install `_CompilerPluginSupport` module directory.

Fix the generated layout for the `_CompilerPluginSupport` Swift module,
so that properly-named Swift interfaces are placed within a
`.swiftmodule` subdirectory. Install that directory into lib/swift.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

rxwei commented on lib/CompilerPluginSupport/CMakeLists.txt:
```diff
@@ -56,4 +77,7 @@ if(SWIFT_SWIFT_PARSER)
       DESTINATION "lib${LLVM_LIBDIR_SUFFIX}"
       COMPONENT compiler)
 
+    swift_install_in_component(DIRECTORY "${module_base}"
```
Do we still need line 66-72?

DougGregor replied:
I think that's installing the dylib itself

DougGregor replied:
They're both installed into `lib`, without the platform, because they're host content.

rxwei replied:
Sorry I meant to say `lib/swift`. Currently the library is installed to `lib` but the module is installed to "lib${LLVM_LIBDIR_SUFFIX}/swift".

DougGregor replied:
Both are going into `lib` with this pull request. We could decide that they go somewhere else, but there's some precedent for using `lib`: `lib_InternalSwiftScan.dylib`, `lib_InternalSwiftStaticMirror.dylib`, `libsourcekitdInProc.dylib` and `libswiftDemangle.dylib`, all host libraries, all go there.

rxwei replied:
`lib` makes sense to me.

DougGregor committed:
Ensure that the compiler links against the plugin support library.

DougGregor committed:
Link _CompilerPluginSupport into the compiler and SourceKit.

Ensure that we link `swift_CompilerPluginSupport` into the compiler and
SourceKit, and set the rpath appropriately to find the library in its
installed location.

A number of driver tests copy the driver executable into a temporary
directory to isolate it from the build tree. Also copy the plugin
support library into its appropriate place near the driver executable
to ensure these tests keep working. To help with this, add a
`swift_swift_parser` lit feature, which we can use in tests that
involve the new parser's capabilities.

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
[Macros] Add a flag `-dump-macro-expansions` as a debugging aid.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test macOS

DougGregor committed:
[Macros] Use appropriate substitutions in tests

DougGregor committed:
[CMake] Fixup build for compiler plugin support library

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
[CMake] Use "rm -f" when removing old Swift module.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
[Macro tests] Pass appropriate linker flags for macro plugin tests.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
LLDB failure looks unrelated:
  lldb-api :: lang/swift/clangimporter/static_archive/TestSwiftStaticArchiveTwoSwiftmodules.py

rerunning

DougGregor commented:
@swift-ci please smoke test macOS

DougGregor commented:
@swift-ci please smoke test macOS

DougGregor committed:
[SourceKit] Make sure we link the compiler plugin support library

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
Huh:

```
Exit Code: -9
Timeout: Reached timeout of 600 seconds

Command Output (stdout):
--
lldb version 13.0.0git (https://github.com/apple/llvm-project.git revision b2590c88ba87f8a3a3130cba4e235dfebdbb4cff)
Apple Swift version 5.8-dev (LLVM b2590c88ba87f8a, Swift 7cb5875ef98af4d)

--
Command Output (stderr):
--
warning: Swift error in scratch context: New Swift image added: /Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/build/buildbot_incremental/lldb-macosx-x86_64/lldb-test-build.noindex/lang/swift/struct_change_rerun/TestSwiftStructChangeRerun.test_swift_struct_change_rerun_dsym/a.out. ClangImporter needs to be reinitialized..
Shared Swift state for a.out has developed fatal errors and is being discarded.
REPL definitions and persistent names/types will be lost.
```

DougGregor commented:
@swift-ci please smoke test macOS
